src/vision/status_quo/alan_needs_async_in_traits.md

```diff
@@ -0,0 +1,101 @@
+    async fn perform(request: Request) -> Response;
+}
+```
+
+When Alan tries to compile this, he gets an error:
+
+```ignore
+ --> src/lib.rs:2:5
+  |
+2 |     async fn perform(request: Request) -> Response;
+  |     -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  |     |
+  |     `async` because of this
+  |
+  = note: `async` trait functions are not currently supported
+  = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
+```
+
+Alan, using Rust for a little while now, has learned to follow compiler error messages and adds `async-trait` to the dependencies list. Alan follows the README of `async-trait` and gets the code to compile. Alan is somewhat dismayed that the compile times have increased considerably after adding this new dependency.
+
+**TODO::** Maybe add representive pre and post trait compile times here.
+
+After Alan finishes adding the new trait, he shows his work off to Barbara and mentions he's happy with the work but is a little sad that compile times have worsened. Barbara, an experienced Rust developer, knows that using `async-trait` comes with some issues and in this case is especially worried about tying their public API to a third-party dependency. She decides to not worry Alan with this right now. Alan and Barbara are pretty happy with the results and go on to publish their crate which gets lots of users.
+
+Later on, a potential user of the library wants to use their library in a `no_std` context where they will be providing a custom HTTP stack. Alan and Barbara have done a pretty good job of limiting the use of standard library features and think it might be possible to support this use case. However, they quickly run into a show stopper: `async-trait` boxes all of the futures returned from a async trait function. They report this to Alan through an issue.
+
+Alan, feeling (over-) confident in his Rust skills, decides to try to see if he can implement async traits without using `async-trait`. 
+
+```rust 
+struct Request(String);
+
+trait HttpClient {
+   type Response: Future<Output = ()>;
+
+   fn perform(request: Request) -> Self::Response; 
+}
+```
+
+Alan seems to have something working, but when he goes to update the examples of how to implement this trait in his crate's documentation, he realizes that he either needs to:
+
+* use trait object:
+
+  ```rust,ignore
+  struct ClientImpl;
+
+  impl HttpClient for ClientImpl {
+      type Response = Pin<Box<dyn Future<Output = ()>>>;
+
+      fn perform(request: Request) -> Self::Response {
+          Box::pin(async move {
+              println!("{}", request.0);
+          })
+	  }
+  }
+  ```
+
+  which wouldn't work for `no_std`.
+
+* implement `Future` trait manually, which isn't particulary easy/straight-forward for non-trivial cases, especially if it involves making other async calls (likely).
+
+**TODO::** Would make sense for Alan to reach out to Barbara at this point and she should be able to give some good advice.
```

Comment:
There isn't *that* much he can do, right? Maybe I don't know the tricks =)

Reply:
She might suggest writing a custom proc-macro that is like async-trait, but has a special no_std mode which allows using it without std? Alan as a newbie in Rust doesn't know much about proc-macros and how to write them himselfes. It's also a huge project for a Rust beginner, but for Barbara as an advanced Rust Developer, she probably has experience with proc-macros and can help out to write it.

Another tip she might give, is how to implement the `Future` trait properly and how to use / create helpers like the `std::task::ready` macro and other stuff.

Reply:
There is the `poll_x` pattern, and I saw a [cool crate recently that can help with that](https://crates.io/crates/async-trait-ext).... but I guess that falls under implementing the `Future` manually/with a different proc-macro.

Reply:
The only way to make things work with no-std that I know of is to rely on nightly only features, is the thing.

Reply:
I would simply have Barbara mention exactly what Niko said - it's likely not possible without using nightly-only features. We can then have compiling on stable as a part of their requirements, leading this to be a dead end.

Reply:
What I was hoping for here is that some of you might know of some API that **somehow** allows you to turn an `impl Future` into a named future, similar to how [`futures::stream::unfold`](https://docs.rs/futures/0.3.13/futures/stream/fn.unfold.html) allows you to create a stream out of an `impl Future` but I forgot that will [require boxing](https://gitlab.freedesktop.org/dbus/zbus/-/blob/491e5456d66510110ee8e730f3bb9e5a692844c0/zbus/src/azync/proxy.rs#L310) as well.

As for a proc-macro way, from what I understood from the [README](https://crates.io/crates/async-trait-ext), `async-trait-ext` only requires a nightly feature for provided methods in traits. Alan doesn't necessarily need provided methods so he can possibly make use of `async-trait-ext` w/o nightly, can't he?